### PR TITLE
Add Test for Mocking SOAR Server Downtime

### LINKED ISSUE #11: JSON Decode error while service was unavailable
Bit of an edge case and realise this is probably only temporary package but while the SOAR was down for maintenance got a json decode error -

```python
result = Fido.search(Time('2021-01-01', '2021-01-15') & Instrument('EUI'))
---------------------------------------------------------------------------
JSONDecodeError                           Traceback (most recent call last)
<ipython-input-8-1b961ccfb6a9> in <module>
----> 1 result = Fido.search(Time('2021-01-01', '2021-01-15') & Instrument('EUI'))

~/.virtualenvs/radiospectra/lib/python3.7/site-packages/sunpy/net/fido_factory.py in search(self, *query)
    346         """
    347         query = attr.and_(*query)
--> 348         results = query_walker.create(query, self)
    349
    350         # If we have searched the VSO but no results were returned, but another

~/.virtualenvs/radiospectra/lib/python3.7/site-packages/sunpy/net/attr.py in create(self, *args, **kwargs)
    563         Call the create function(s) matching the arguments to this method.
    564         """
--> 565         return self.createmm(self, *args, **kwargs)
    566
    567     def apply(self, *args, **kwargs):

~/.virtualenvs/radiospectra/lib/python3.7/site-packages/sunpy/util/functools.py in wrapper(*args, **kwargs)
     16
     17     def wrapper(*args, **kwargs):
---> 18         return dispatcher.dispatch(args[1].__class__)(*args, **kwargs)
     19
     20     wrapper.dispatch = dispatcher.dispatch

~/.virtualenvs/radiospectra/lib/python3.7/site-packages/sunpy/net/fido_factory.py in _create_and(walker, query, factory)
    275 @query_walker.add_creator(attr.AttrAnd)
    276 def _create_and(walker, query, factory):
--> 277     return factory._make_query_to_client(*query.attrs)
    278
    279

~/.virtualenvs/radiospectra/lib/python3.7/site-packages/sunpy/net/fido_factory.py in _make_query_to_client(self, *query)
    511             if isinstance(tmpclient, vso.VSOClient):
    512                 kwargs = dict(response_format="table")
--> 513             results.append(tmpclient.search(*query, **kwargs))
    514
    515         # This method is called by `search` and the results are fed into a

~/.virtualenvs/radiospectra/lib/python3.7/site-packages/sunpy_soar/client.py in search(self, *query, **kwargs)
     25         results = []
     26         for query_parameters in queries:
---> 27             results.append(self._do_search(query_parameters))
     28         table = astropy.table.vstack(results)
     29         qrt = QueryResponseTable(table, client=self)

~/.virtualenvs/radiospectra/lib/python3.7/site-packages/sunpy_soar/client.py in _do_search(query)
     63
     64         # Do some list/dict wrangling
---> 65         names = [m['name'] for m in r.json()['metadata']]
     66         info = {name: [] for name in names}
     67         for entry in r.json()['data']:

~/.virtualenvs/radiospectra/lib/python3.7/site-packages/requests/models.py in json(self, **kwargs)
    898                     # used.
    899                     pass
--> 900         return complexjson.loads(self.text, **kwargs)
    901
    902     @property

~/.pyenv/versions/3.7.0/Python.framework/Versions/3.7/lib/python3.7/json/__init__.py in loads(s, encoding, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)
    346             parse_int is None and parse_float is None and
    347             parse_constant is None and object_pairs_hook is None and not kw):
--> 348         return _default_decoder.decode(s)
    349     if cls is None:
    350         cls = JSONDecoder

~/.pyenv/versions/3.7.0/Python.framework/Versions/3.7/lib/python3.7/json/decoder.py in decode(self, s, _w)
    335
    336         """
--> 337         obj, end = self.raw_decode(s, idx=_w(s, 0).end())
    338         end = _w(s, end).end()
    339         if end != len(s):

~/.pyenv/versions/3.7.0/Python.framework/Versions/3.7/lib/python3.7/json/decoder.py in raw_decode(self, s, idx)
    353             obj, end = self.scan_once(s, idx)
    354         except StopIteration as err:
--> 355             raise JSONDecodeError("Expecting value", s, err.value) from None
    356         return obj, end

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

### PR DESCRIPTION
**Pull request by NucleonGodX**
Adding a test case to mimic SOAR server downtime resulting in JSON decode error

Fixes #11

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -20,6 +20,7 @@ dependencies = [
   "astropy>5.2.0",
   "sunpy[net]>=5.0",
   "requests>=2.28.0",
+  "responses>=0.20.0"
```
Comment (nabobalis) [inline pyproject.toml:23]: What is this library? 

It also should be moved to be under the tests group.

File: sunpy_soar/tests/test_sunpy_soar.py [hunk 2]
```diff
@@ -268,3 +270,21 @@ def test_join_low_latency_query():
         "+WHERE+h1.instrument='EUI'+AND+h1.begin_time>='2021-02-01+00:00:00'+AND+h1.begin_time<='2021-02-02+00:00:00'"
         "+AND+h2.dimension_index='1'+AND+h1.level='LL01'+AND+h1.descriptor='eui-fsi174-image'"
     )
+
+
+@pytest.mark.xfail(reason="Expected to fail due to invalid JSON response causing JSONDecodeError")
```
Comment (nabobalis) [inline sunpy_soar/tests/test_sunpy_soar.py:275]: The test should check the error is raised with `pytest.raises`, not be marked xfail. 

See https://docs.pytest.org/en/7.1.x/how-to/assert.html

File: pyproject.toml (hunk 1 above)
Comment (NucleonGodX) [reply pyproject.toml:23]: Its a library used to simulate different HTTP responses without making actual network calls. It allows one to control the responses and ensure that the code behaves correctly under different test conditions.
https://pypi.org/project/responses/

File: sunpy_soar/tests/test_sunpy_soar.py (hunk 2 above)
Comment (NucleonGodX) [reply sunpy_soar/tests/test_sunpy_soar.py:275]: Alrighty.

File: pyproject.toml (hunk 1 above)
Comment (nabobalis) [reply pyproject.toml]: Looks very nice

File: sunpy_soar/tests/test_sunpy_soar.py [hunk 3]
```diff
@@ -268,3 +271,22 @@ def test_join_low_latency_query():
         "+WHERE+h1.instrument='EUI'+AND+h1.begin_time>='2021-02-01+00:00:00'+AND+h1.begin_time<='2021-02-02+00:00:00'"
         "+AND+h2.dimension_index='1'+AND+h1.level='LL01'+AND+h1.descriptor='eui-fsi174-image'"
     )
+
+
+@responses.activate
+def test_SOAR_server_down():
```
Comment (nabobalis) [inline sunpy_soar/tests/test_sunpy_soar.py:277]:
```suggestion
def test_soar_server_down():
```

File: sunpy_soar/tests/test_sunpy_soar.py [hunk 4]
```diff
@@ -268,3 +271,22 @@ def test_join_low_latency_query():
...
+    )
+    # We do not give any json data similar to the condition when the server is down.
+    responses.add(responses.GET, TAP_ENDPOINT, body="Invalid JSON response", status=200)
+
+    time = a.Time("2020-11-13", "2020-11-14")
+    level = a.Level("LL02")
+    product = a.soar.Product("mag")
+
+    with pytest.raises(JSONDecodeError):
+        Fido.search(time, level, product)
```
Comment (nabobalis) [inline sunpy_soar/tests/test_sunpy_soar.py:292]: Where is the json error raised? Is it something we can capture in the soar client?

File: tox.ini [hunk 5]
```diff
@@ -34,6 +34,7 @@ set_env =
 deps =
     # For packages which publish nightly wheels this will pull the latest nightly
     devdeps: astropy>=0.0.dev0
+    devdeps: responses
```
Comment (nabobalis) [inline tox.ini:37]:
```suggestion
```

File: sunpy_soar/tests/test_sunpy_soar.py (hunk 4 above)
Comment (NucleonGodX) [reply sunpy_soar/tests/test_sunpy_soar.py:292]: I think so that can be done.

File: sunpy_soar/client.py [hunk 6]
```diff
@@ -170,12 +171,17 @@ def _do_search(query):
         r = requests.get(f"{tap_endpoint}/sync", params=payload)
         log.debug(f"Sent query: {r.url}")
         r.raise_for_status()
+        try:
+            response_json = r.json()
+        except JSONDecodeError as err:
+            msg = "Server returned an invalid JSON response. The SOAR server may be down or not functioning correctly."
```
Comment (nabobalis) [inline sunpy_soar/client.py:177]: So in the original case, the error was up and returned a 200 but failed to provide any json?

Comment (nabobalis) [inline sunpy_soar/client.py:177]:
```suggestion
            msg = (
                "The SOAR server returned an invalid JSON response. "
                "It may be down or not functioning correctly."
            )
```
Not sure if the python whitespace is correct.

**Review by nabobalis — APPROVED** [review]

File: sunpy_soar/client.py (hunk 6 above)
Comment (nabobalis) [reply sunpy_soar/client.py:177]: The test match will need to be updated.

Comment (NucleonGodX) [reply sunpy_soar/client.py:177]: Yep

**Comment by nabobalis** [comment]
Thanks @NucleonGodX
